File: adr_generate.py
#!/usr/bin/env python3

import os
import sys
import openai
import argparse
from openai import OpenAI

def read_input(file_path=None):
    """
    Reads input from a file if a file path is provided, otherwise reads from standard input.

    Args:
        file_path (str, optional): The path to the file to read from. If not provided, reads from standard input.

    Returns:
        str: The content read from the file or standard input.

    Raises:
        FileNotFoundError: If the specified file does not exist.
    """
    if file_path:
        try:
            with open(file_path, 'r') as file:
                return file.read()
        except FileNotFoundError:
            stop_with_error(f"File {file_path} not found.",1)
    else:
        say("Reading from stdin")
        return sys.stdin.read()

def read_prompt(prompt_file):
    """
    Reads the content of a prompt file.

    Args:
        prompt_file (str): The name of the prompt file to read.

    Returns:
        str: The content of the prompt file.

    Raises:
        FileNotFoundError: If the prompt file does not exist.

    Note:
        The function constructs the full path to the prompt file based on the
        directory of the current script.
    """
    script_dir = os.path.dirname(os.path.realpath(__file__))
    prompt_path = os.path.join(script_dir, prompt_file)
    try:
        with open(prompt_path, 'r') as file:
            return file.read()
    except FileNotFoundError:
        stop_with_error(f"Prompt file {prompt_path} not found.",2)

def generate_adr(design_text, prompt, model, api_key):
    """
    Generates an Architecture Decision Record (ADR) using the OpenAI API.
    Args:
        design_text (str): The design text to be included in the ADR.
        prompt (str): The prompt to guide the OpenAI model.
        model (str): The OpenAI model to use for generating the ADR.
        api_key (str): The API key for authenticating with the OpenAI API.
    Returns:
        str: The generated ADR content.
    Raises:
        SystemExit: If the API key is not provided or if there is an error with the OpenAI API.
    """
    openai.api_key = api_key
    
    if openai.api_key is None:
        stop_with_error("Error: Please provide an OpenAI API key either through the command line or environment variable.",3)


    try:
        client = OpenAI(api_key=api_key)
        response = client.chat.completions.create(  model=model,
                                                    messages=[
                                                        {"role": "system", "content": prompt},
                                                        {"role": "user", "content": design_text}
                                                    ],
                                                    max_tokens=500,
                                                    temperature=0.7)

        return response.choices[0].message.content.strip()
    
    except openai.OpenAIError as e:
        stop_with_error(f"Error with OpenAI API: {str(e)}",4)


def write_output(output_text, output_file):
    """
    Writes the given text to the specified output file.

    Args:
        output_text (str): The text to be written to the file.
        output_file (str): The path to the file where the text will be written.

    Raises:
        IOError: If there is an error writing to the file.
    """
    try:
        with open(output_file, 'w') as file:
            file.write(output_text)
    except IOError:
        stop_with_error(f"Error writing to {output_file}",5)

def stop_with_error(msg,code=1):
    """
    Print an error message and exit the program.
    """
    print(msg,file=sys.stderr)
    sys.exit(code)

def parse_arguments():
    """
    Parse CLI arguments using argparse.
    """
    parser = argparse.ArgumentParser(description="Generate an Architecture Decision Record (ADR) using OpenAI Model.")
    
    # Define the optional parameters
    parser.add_argument('-i', '--input', type=str, help="Input text file with design discussion", default=None)
    parser.add_argument('-p', '--prompt', type=str, help="Prompt file to use", default="prompt.txt")
    parser.add_argument('-o', '--output', type=str, help="Output file for the ADR", default="adr.md")
    parser.add_argument('-m', '--model', type=str, help="OpenAI model to use", default="gpt-4")
    parser.add_argument('-k', '--api_key', type=str, help="OpenAI API key", default=os.getenv("OPENAI_API_KEY"))
    
    return parser.parse_args()

def say(message):
    """
    Print the given message.
    """
    print(message)

def main():

    args = parse_arguments()

    design_text = read_input(args.input)
    
    prompt = read_prompt(args.prompt)

    say(f"Generating ADR using model {args.model}")
    adr = generate_adr(design_text, prompt, args.model, args.api_key)

    say(f"Writing output to {args.output}")
    write_output(adr, args.output)

    say("Done!")
   

if __name__ == "__main__":
    main()

File: adr-generate.js
const { Command } = require('commander');
const fs = require('fs').promises;
const axios = require('axios');

const BASE_URL = "https://cohost.gue5ty.com";
const GUEST_OPENWEB_UI_URL = `${BASE_URL}/api/chat/completions`;
const NEW_CHAT_URL = `${BASE_URL}/api/v1/chats/new`;
const DELETE_CHAT_URL = `${BASE_URL}/api/v1/chats/{ID}`
const apiKey = process.env.OPENWEBUI_API_KEY;
// const MODEL = 'claude-sonnet-35';
const MODEL = 'aws_bedrock_claude_pipeline.anthropic.claude-3-5-sonnet-20241022-v2:0';
const FIRST_CHOICE = 0;
const DEFAULT_PROMPT_TEMPLATE = 'adr-prompt.txt';

const DEFAULT_REQUEST_OPTIONS = {
    headers: {
        'Content-Type': 'application/json',
        'Authorization': `Bearer ${apiKey}`
    }
};

const ERR_CREATE_CHAT = 1;
const ERR_DELETE_CHAT = 2;
const ERR_GET_CHAT_RESPONSE = 3;
const ERR_UNEXPECTED = 4;
const ERR_READ_FILE = 5;

function readFile(fileName) {
    return fs.readFile(fileName, 'utf8')
        .catch(error => {
            console.error(`Error: File '${fileName}' not found.`);
            process.exit(ERR_READ_FILE);
        });
}

function createPrompt(discussion, promptTemplate) {
    
    const adrPrompt = promptTemplate.replace('{discussion}', discussion);
    return adrPrompt;
}

async function createNewChat()
{
    try
    {
        const newChatPayload =
        {
            "chat" : {
                "id" : "",
                "title" : "new title",
                "model": MODEL,
                "messages": []
            }
        }
        const response = await axios.post(NEW_CHAT_URL, newChatPayload, DEFAULT_REQUEST_OPTIONS);
        return response.data.id; // Assuming the API returns the chat ID in the 'id' field
    }
    catch (error)
    {
        console.error(`Error creating new chat: ${error.message}`);
        process.exit(ERR_CREATE_CHAT);
    }
}

async function deleteChat(id)
{
    try
    {
        let url = DELETE_CHAT_URL.replace('{ID}',id)
        const response = await axios.delete(url, DEFAULT_REQUEST_OPTIONS)
        return response.data
    }
    catch (err)
    {
        console.error(`Error deleting chat ${err.message}`)
        process.exit(ERR_DELETE_CHAT);
    }
}

async function getResponseFromOpenWebUI(prompt,chatID)
{
    
    const requestData =
    {
        chat_id : chatID,
        model: MODEL,
        stream : false,
        messages: [{ role: "user", content: prompt }]
    };
    
    try
    {
        const response = await axios.post(GUEST_OPENWEB_UI_URL, requestData, DEFAULT_REQUEST_OPTIONS);
        return response.data.choices[FIRST_CHOICE].message.content;
    }
    catch (error)
    {
        console.error(`Error communicating with OpenWebUI: ${error.message}`);
        process.exit(ERR_GET_CHAT_RESPONSE);
    }
}

async function readStdin()
{
    return new Promise((resolve, reject) => {
        let data = '';
        process.stdin.setEncoding('utf8');

        process.stdin.on('readable', () => {
            let chunk;
            while ((chunk = process.stdin.read()) !== null) {
                data += chunk;
            }
        });

        process.stdin.on('end', () => {
            resolve(data);
        });

        process.stdin.on('error', (err) => {
            reject(err);
        });
    });
}

async function main()
{
    const options = parseInputArgs();

    let chatID = '';
    try
    {
        
        let inputContent = await getInputText(options.input);
        
        let prompt = await getPromptFromFile(options.prompt,inputContent);
        
        chatID = await createNewChat();
        const response = await getResponseFromOpenWebUI(prompt, chatID);
        
        await outputResult(options.output, response);
    }
    catch (error)
    {
        throw error;
    }
    finally
    {
        if (chatID)
            await deleteChat(chatID);
    }
}

main().catch(error => {
    console.error(`Unexpected error: ${error.message}`);
    process.exit(ERR_UNEXPECTED);
});

async function outputResult(outputFile, response)
{
    if (outputFile) {
        await fs.writeFile(outputFile, response, 'utf8');
    } else {
        console.log(response);
    }
}

async function getPromptFromFile(fileName,inputContent)
{
    if (!fileName) { throw new Error('Prompt file is required'); }
    const promptTemplate = await readFile(fileName);
    let prompt = createPrompt(inputContent, promptTemplate);
    return prompt;
}

async function getInputText(input)
{
    return input ? await readFile(input) : await readStdin();
}

function parseInputArgs()
{
    const program = new Command();

    program
        .name('adr-generate')
        .description('Generate Architecture Decision Records (ADRs) from discussions using AI')
        .version('1.0.0')
        .option('-i, --input <file>', 'input file (defaults to reading from stdin)', null)
        .option('-o, --output <file>', 'output file (defaults to writing to stdout)', null)
        .option('-p, --prompt <file>', 'prompt file', 'default_claude_adr_prompt.txt');

    program.parse(process.argv);

    const options = program.opts();
    return options;
}
